zephyr/lib/regions_mm.c

```diff
@@ -1,54 +1,574 @@
+			    SOF_DIV_ROUND_UP(
+				cfg->block_bundles_table[i].number_of_blocks,
+				8),
+			    sizeof(uint32_t));
+
+			uint32_t *bitaray_bitfield;
+
+			bitaray_bitfield =
+			    rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				    sizeof(uint32_t) * bitaray_size);
+
+			if (bitaray_bitfield == NULL)
+				goto fail;
+
+			allocators_bitaray->num_bits =
+			    cfg->block_bundles_table[i].number_of_blocks;
+			allocators_bitaray->num_bundles = bitaray_size;
+			allocators_bitaray->bundles = bitaray_bitfield;
+
+			new_allocator->num_blocks =
+			    cfg->block_bundles_table[i].number_of_blocks;
+			new_allocator->blk_sz_shift =
+			    ilog2(cfg->block_bundles_table[i].block_size);
+			new_allocator->buffer =
+			    (uint8_t *)new_heap->virtual_region->addr + offset;
+			new_allocator->bitmap = allocators_bitaray;
+
+			offset += cfg->block_bundles_table[i].number_of_blocks *
+				  cfg->block_bundles_table[i].block_size;
+			new_heap->physical_blocks_allocators[i] = new_allocator;
+		}
+	}
+
+	new_heap->span = span;
+
+	/* Save the new heap pointer to a linked list */
+	list_item_append(&vm_heaps, &new_heap->node);
+
+	return new_heap;
+fail:
+	for (int i = 0; i < MAX_MEMORY_ALLOCATORS_COUNT; i++) {
+		struct sys_mem_blocks *allocator =
+		    new_heap->physical_blocks_allocators[i];
+			rfree(allocator->bitmap);
+			rfree(allocator);
+	}
+	rfree(new_heap);
+	return NULL;
+}
 
 /**
- * @brief Fill vm_heaps array with information from zephyr.
+ * @brief Alloc function
+ *
+ * Allocates memory on heap from provided heap pointer.
+ * Check if we need to map physical memory for given allocation
+ * and do it when nesscesary.
+ * @param heap pointer to a heap that will be used for allocation.
+ * @param alloc_size Size in bytes to allocate.
  *
- * Virtual memory regions calculated in zephyr are translated here
- * to a struct that will keep all information on current allocations
- * and virtual to physical mappings that are related to heaps.
- * System heap is not a part of this information. It only refers to
- * virtual first heaps.
- * Has to be initialized after calculations for regions is done in zephyr.
+ * @retval ptr to a new heap if successful
+ * @retval NULL on allocation failure
  */
-static int virtual_heaps_init(const struct device *unused)
+void *vmh_alloc(struct virtual_memory_heap *heap, uint32_t alloc_size)
 {
-	ARG_UNUSED(unused);
+	void *ptr = NULL;
+	int mem_block_iterator;
+	uint32_t block_count;
 
-	struct sys_mm_drv_region *virtual_memory_regions =
-		(struct sys_mm_drv_region *)sys_mm_drv_query_memory_regions();
+	/* Error code we will gather when allocating on physical block
+	 * allocators.
+	 * Loop will continiue as long as there are blocks to iterate on and
+	 * if allocation keeps returning error code.
+	 * When alloc succeeded it will return 0 on allocation and we break
+	 * form loop and continiue.
+	 */
+	int allocation_error_code = -ENOMEM;
+	struct vmh_active_allocation *new_allocation =
```

Comment:
What is wrong with `ret` or `err` instead of `allocation_error_code`?

Reply:
I like my variables to be meaningful I find it more understandable if I know where the error code is coming from. Especially if there is code is longer and it is not returning error of the function it is in..

Reply:
@dabekjakub yes, but this is an open source project and you need to adhere to existing practices. This way the code can be easily maintained, and people can quickly review your code.

Linux kernel has explicitly set a limit to 100 chars per line to avoid this kind of situations. 

For example instead of 3 lines like this:

```
struct virtual_memory_heap *new_heap =
	    rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
		    sizeof(struct virtual_memory_heap));
```

you could have easily fitted 100 chars with:

```
struct vmh_heap *nh = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*nh);
```

I understand that this is how you used to write your code in your previous projects, but we need to understand that this code needs to be maintained. And the easiest way for that is to write it in a style that is already existing in the project.

Since most of the people here come from a Linux kernel background we use:

https://www.kernel.org/doc/html/v4.10/process/coding-style.html

in a pretty relaxed way.

Reply:
While that's a thing when you're doing this on its own, in the greater project such an API would be more confusing than useful. You're writing in essence a memory allocator, so the returns should behave like those of any memory allocator.

It is clear even with the short name that it's an allocation error return -- what else could it possibly be in a function called vmh_alloc?

Reply:
@dabekjakub @paulstelian97 Let's not start a war over a variable name (yet!).

Please fix the checkpatch.pl warnings and we are good to merge. You can re-arrange the code later. 

```
CHECK: Lines should not end with a '('
#264: FILE: zephyr/include/sof/lib/regions_mm.h:223:
+			struct vmh_active_physical_mapping *mapping = CONTAINER_OF(

CHECK: Lines should not end with a '('
#267: FILE: zephyr/include/sof/lib/regions_mm.h:226:
+			need_mapping = vmh_is_ptr_in_memory_range(

CHECK: Comparison to NULL could be written "vmh_get_heap_by_attribute"
#319: FILE: zephyr/lib/regions_mm.c:31:
+	if (vmh_get_heap_by_attribute(memory_region_attribute, core_id) != NULL)

CHECK: Comparison to NULL could be written "!new_heap"
#325: FILE: zephyr/lib/regions_mm.c:37:
+	if (new_heap == NULL)

CHECK: Blank lines aren't necessary after an open brace '{'
#343: FILE: zephyr/lib/regions_mm.c:55:
+		for (i = CONFIG_MP_MAX_NUM_CPUS; i < VIRTUAL_REGION_COUNT; i++) {
+

CHECK: Comparison to NULL could be written "!cfg"
#352: FILE: zephyr/lib/regions_mm.c:64:
+	if (cfg == NULL) {

CHECK: Lines should not end with a '('
#386: FILE: zephyr/lib/regions_mm.c:98:
+			struct sys_mem_blocks *new_allocator = rzalloc(

CHECK: Comparison to NULL could be written "!new_allocator"
#389: FILE: zephyr/lib/regions_mm.c:101:
+			if (new_allocator == NULL)

CHECK: Lines should not end with a '('
#392: FILE: zephyr/lib/regions_mm.c:104:
+			struct sys_bitarray *allocators_bitaray = rzalloc(

CHECK: Comparison to NULL could be written "!allocators_bitaray"
#395: FILE: zephyr/lib/regions_mm.c:107:
+			if (allocators_bitaray == NULL) {

CHECK: Lines should not end with a '('
#401: FILE: zephyr/lib/regions_mm.c:113:
+			size_t bitaray_size = SOF_DIV_ROUND_UP(

CHECK: Comparison to NULL could be written "!bitaray_bitfield"
#410: FILE: zephyr/lib/regions_mm.c:122:
+			if (bitaray_bitfield == NULL)

CHECK: Lines should not end with a '('
#485: FILE: zephyr/lib/regions_mm.c:186:
+	struct vmh_active_allocation *new_allocation = rzalloc(

CHECK: Comparison to NULL could be written "!new_allocation"
#488: FILE: zephyr/lib/regions_mm.c:189:
+	if (new_allocation == NULL)

CHECK: Lines should not end with a '('
#514: FILE: zephyr/lib/regions_mm.c:215:
+			allocation_error_code = sys_mem_blocks_alloc_contiguous(

CHECK: Lines should not end with a '('
#518: FILE: zephyr/lib/regions_mm.c:219:
+			allocation_error_code = sys_mem_blocks_alloc(

CHECK: Comparison to NULL could be written "!ptr"
#536: FILE: zephyr/lib/regions_mm.c:237:
+	if (ptr == NULL) {

CHECK: Blank lines aren't necessary after an open brace '{'
#575: FILE: zephyr/lib/regions_mm.c:276:
+			if (ptrs_to_map[i] != 0) {
+

CHECK: Lines should not end with a '('
#600: FILE: zephyr/lib/regions_mm.c:301:
+		sys_mem_blocks_free_contiguous(

CHECK: Blank lines aren't necessary after an open brace '{'
#665: FILE: zephyr/lib/regions_mm.c:358:
+		    CONTAINER_OF(iterator, struct vmh_active_allocation, node)->allocation_ptr) {
+

CHECK: Comparison to NULL could be written "!allocation"
#673: FILE: zephyr/lib/regions_mm.c:364:
+	if (allocation == NULL)

CHECK: Lines should not end with a '('
#679: FILE: zephyr/lib/regions_mm.c:370:
+		if (vmh_is_ptr_in_memory_range(

CHECK: Lines should not end with a '('
#684: FILE: zephyr/lib/regions_mm.c:375:
+				retval = sys_mem_blocks_free_contiguous(

CHECK: Lines should not end with a '('
#720: FILE: zephyr/lib/regions_mm.c:409:
+			if (vmh_do_ranges_overlap(

CHECK: Blank lines aren't necessary after an open brace '{'
#723: FILE: zephyr/lib/regions_mm.c:412:
+				physical_maping->mapping_ptr, physical_maping->mapping_size)) {
+

```

Reply:
@paulstelian97 in this specific example I should have said that there will be telemetry added here in future so it will be for distinction between allocation and mapping errors. And still I find it just more understandable than constant over abbreviating everything.

This discussion aside i made changes according to checkpatch in places where it was applicable. Also I have some kind of false positives in main one - will push and see the result maybe its my local one.

Reply:
Yep it shows up as over 100 lines and it is not in every editor I have. Its either that or open brace at begining of line.

Reply:
Telemetry errors shouldn't propagate to the caller. Eh, I suppose this can stay like this but it still feels fishy despite that. Focus on the hard errors from checkpatch, we don't want to merge anything that does that.

Reply:
The last hard error is a false positive